# Phase 2: Data Cleaning

Summary
---
* **Section 1: Importing libraries & the datasets**
---
* **Section 2: Exploring & understanding each dataset**
    1. Explored departments.csv
        * Checked for null & duplicate values. Found none.
    <br><br>
    2. Explored aisles.csv
        * Checked for null & duplicate values. Found none.
    <br><br>
    3. Explored products.csv
        * Checked for null & duplicate values. Found none.
    <br><br>
    4. Explored orders.csv
        * Made sure that all order_ids present in the "order_id" column are unique
        * Checked for null & duplicate values. Null values were present and were dealt with accordingly
        * Investigated the total number of users in the dataset.
        * Checked whether column "order_dow" is in range from 0 - 6. Also checked whether column "order_hour_of_day" is in range from 0 - 23
        * Removed data belonging to train and test set. This was identified using the "eval_set" column.
        * Saved the cleaned dataset.
    <br><br>
    5. Explored order_products__prior.csv
        * Checked for null values, found none. Did not check for duplicate values because they are expected to  be present in the dataset.
        * Checked whether the order_ids present in this dataset are the same as the order_ids present in the "orders.csv" dataset
        * Checked whether "reordered" column contained only binary data.

## ---- Importing the libraries and datasets ----

In [1]:
import pandas as pd
import numpy as np

In [2]:
aisles_df = pd.read_csv('aisles.csv')
depts_df = pd.read_csv('departments.csv')
order_prod_df = pd.read_csv('order_products__prior.csv')
orders_df = pd.read_csv('orders.csv')
products_df = pd.read_csv('products.csv')

## ---- Exploring and understanding each dataset ----

#### xxxxxxxxx Exploring departments.csv xxxxxxxxx

In [18]:
print(depts_df.shape)
depts_df.head()

(21, 2)


department_id department
0              1     frozen
1              2      other
2              3     bakery
3              4    produce
4              5    alcohol

In [46]:
depts_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21 entries, 0 to 20
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   department_id  21 non-null     int64 
 1   department     21 non-null     object
dtypes: int64(1), object(1)
memory usage: 464.0+ bytes


Now we need to check whether there are any null values present or any duplicate values present in the dataset.

In [30]:
# Checking null
depts_df.isnull().any()

department_id    False
department       False
dtype: bool

In [24]:
# Checking duplicates
depts_df.duplicated(['department_id','department']).sum()

0

There are no null & duplicate values present in the csv file.
____
____

#### xxxxxxxxx Exploring asiles.csv xxxxxxxxx

In [33]:
print(aisles_df.shape)
aisles_df.head()

(134, 2)


aisle_id                       aisle
0         1       prepared soups salads
1         2           specialty cheeses
2         3         energy granola bars
3         4               instant foods
4         5  marinades meat preparation

In [47]:
aisles_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 134 entries, 0 to 133
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   aisle_id  134 non-null    int64 
 1   aisle     134 non-null    object
dtypes: int64(1), object(1)
memory usage: 2.2+ KB


Now we need to check whether there are any null values present or any duplicate values present in the dataset.

In [35]:
# Checking null
aisles_df.isnull().any()

aisle_id    False
aisle       False
dtype: bool

In [37]:
# Checking duplicates
aisles_df.duplicated(['aisle_id','aisle']).sum()

0

There are no null & duplicate values present in the csv file.
____
____

#### xxxxxxxxx Exploring products.csv xxxxxxxxx

In [38]:
print(products_df.shape)
products_df.head()

(49688, 4)


product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  
0             19  
1             13  
2              7  
3              1  
4             13

In [48]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49688 entries, 0 to 49687
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   product_id     49688 non-null  int64 
 1   product_name   49688 non-null  object
 2   aisle_id       49688 non-null  int64 
 3   department_id  49688 non-null  int64 
dtypes: int64(3), object(1)
memory usage: 1.5+ MB


Now we need to check whether there are any null values present or any duplicate values present in the dataset.

In [40]:
# Checking null
products_df.isnull().any()

product_id       False
product_name     False
aisle_id         False
department_id    False
dtype: bool

In [52]:
# Checking duplicates
products_df.duplicated(['product_id','product_name']).sum()

0

There are no null & duplicate values present in the csv file.
____
____

#### xxxxxxxxx Exploring orders.csv xxxxxxxxx

In [30]:
print(orders_df.shape)
orders_df.head()

(3421083, 7)


order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     0.0  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [4]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   eval_set                object 
 3   order_number            int64  
 4   order_dow               int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(5), object(1)
memory usage: 182.7+ MB


We need to make sure that all order ids present in the order_id column must be unique.

In [5]:
if orders_df.order_id.nunique() == orders_df.shape[0]:
    print('Order_id col is unique')
else:
    print('Data redundancy is present in Order_id col')

Order_id col is unique


Okay now that order_id col is unique, we can move on to check for any null values

In [6]:
orders_df.isnull().any()

order_id                  False
user_id                   False
eval_set                  False
order_number              False
order_dow                 False
order_hour_of_day         False
days_since_prior_order     True
dtype: bool

In [7]:
orders_df.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There seem to be 206209 null values to be present in the days_since_prior_order column. Lets explore what these null values are

In [8]:
null_df = orders_df[orders_df.days_since_prior_order.isnull()]

In [28]:
print(null_df.shape)
null_df.head()

(206209, 7)


order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0    2539329        1    prior             1          2                  8   
11   2168274        2    prior             1          2                 11   
26   1374495        3    prior             1          1                 14   
39   3343014        4    prior             1          6                 11   
45   2717275        5    prior             1          3                 12   

    days_since_prior_order  
0                      NaN  
11                     NaN  
26                     NaN  
39                     NaN  
45                     NaN

In [16]:
null_df.order_number.unique()

array([1])

Every user_id has an order id associated with it. The days_since_prior_order shows the interval between previous order and current order. Every first order placed by a user, surely there will be no data present for previous order as it does not exist.

We confirmed this by checking the unique values present in the order_number col which represents the order number of a user.
___
We will now replace all the null values with zero

In [21]:
orders_df.days_since_prior_order.fillna(0,inplace=True)

In [23]:
orders_df.isnull().any()

order_id                  False
user_id                   False
eval_set                  False
order_number              False
order_dow                 False
order_hour_of_day         False
days_since_prior_order    False
dtype: bool

Let's check how may users are present in the dataset

In [25]:
orders_df.user_id.nunique()

206209

There are approx 2 Lakh (0.2 Million) users available in the dataset

Now we need to make sure that values in col **order_dow range from 0 - 6** and values in col **order_hour_of_day range from  0 - 23**

In [39]:
sorted(orders_df.order_dow.unique())

[0, 1, 2, 3, 4, 5, 6]

In [55]:
orders_df.order_hour_of_day.unique()

array([ 8,  7, 12, 15,  9, 14, 16, 11, 10, 19, 18, 17, 13, 20,  0, 21, 22,
        5, 23,  4,  6,  1,  2,  3])

Both the columns do have the values in intended range.
___
Moving on, we will now explore and clean the eval_set col.

In [56]:
orders_df.eval_set.unique()

array(['prior', 'train', 'test'], dtype=object)

In [61]:
# ----- Dropping rows for train set ----
print(orders_df[orders_df.eval_set == 'train'].shape)
temp = orders_df[orders_df.eval_set == 'train'].index

orders_df.drop(temp,inplace = True)

(131209, 7)


In [68]:
# ----- Dropping rows for test set ----
print(orders_df[orders_df.eval_set == 'test'].shape)
temp = orders_df[orders_df.eval_set == 'test'].index

orders_df.drop(temp,inplace = True)

(75000, 7)


In [72]:
# Now to check whether the previous operation was successful or not
if 3421083 - (131209 + 75000) == orders_df.shape[0]:
    print('Operation was successful')

print(orders_df.eval_set.unique())

Operation was successful
['prior']


In the above operation, we removed all the rows that were associated with test set and train set as there was no need for that data. We checked whether our operation was successfull or not by checking the total number of rows before and after operation and also by checking the unique values present in the eval_set col.

In [79]:
# Saving the cleaned dataset
orders_df.to_csv('cleaned_orders.csv')

In [83]:
# Checking whether the file was properly saved
temp_df = pd.read_csv('cleaned_orders.csv')

In [85]:
print(temp_df.shape)
temp_df.head()

(3214874, 8)


Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0           0   2539329        1    prior             1          2   
1           1   2398795        1    prior             2          3   
2           2    473747        1    prior             3          3   
3           3   2254736        1    prior             4          4   
4           4    431534        1    prior             5          4   

   order_hour_of_day  days_since_prior_order  
0                  8                     0.0  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [86]:
if 3421083 - (131209 + 75000) == temp_df.shape[0]:
    print('Operation was successful')

Operation was successful


The cleaned dataset was successfully saved
___
___

#### xxxxxxxxx Exploring order_products__prior.csv xxxxxxxxx

In [73]:
print(order_prod_df.shape)
order_prod_df.head()

(32434489, 4)


order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [75]:
order_prod_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32434489 entries, 0 to 32434488
Data columns (total 4 columns):
 #   Column             Dtype
---  ------             -----
 0   order_id           int64
 1   product_id         int64
 2   add_to_cart_order  int64
 3   reordered          int64
dtypes: int64(4)
memory usage: 989.8 MB


Lets check for null values

In [74]:
order_prod_df.isnull().any()

order_id             False
product_id           False
add_to_cart_order    False
reordered            False
dtype: bool

There are no null values. Here, there's no need to check for dupicates as it is expected to have multiple duplicate values in the entire dataset.
___
We have to check whether the unique total of order_id is equal to total number of rows / order ids present in orders.csv dataset or not.

In [77]:
if order_prod_df.order_id.nunique() == orders_df.shape[0]:
    print('True')
else:
    print('False')

True


We performed this operation to make sure that unique total of order ids in **order_prod_df** are same as rows / order_ids present in **orders.csv**. If it wasn't equal, then there may have been missing or irrelevant data present in the dataset.

Finally, checking whether reordered col has binary unique values or not.

In [78]:
order_prod_df.reordered.unique()

array([1, 0])

The reordered col has proper binary values as expected.